tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/TripleSourceBase.java

```diff
@@ -138,27 +138,32 @@ public CloseableIteration<BindingSet, QueryEvaluationException> getStatements(
 			// evaluate the query
 			monitorRemoteRequest();
 			CloseableIteration<BindingSet, QueryEvaluationException> res = query.evaluate();
-			resultHolder.set(res);
-
-			// apply filter and/or insert original bindings
-			if (filterExpr != null) {
-				if (bindings.size() > 0) {
-					res = new FilteringInsertBindingsIteration(filterExpr, bindings, res,
-							queryInfo.getStrategy());
-				} else {
-					res = new FilteringIteration(filterExpr, res, queryInfo.getStrategy());
-				}
-				if (!res.hasNext()) {
-					Iterations.closeCloseable(res);
-					conn.close();
-					resultHolder.set(new EmptyIteration<>());
-					return;
+			try {
+				resultHolder.set(res);
+
+				// apply filter and/or insert original bindings
+				if (filterExpr != null) {
+					if (bindings.size() > 0) {
+						res = new FilteringInsertBindingsIteration(filterExpr, bindings, res,
+								queryInfo.getStrategy());
+					} else {
+						res = new FilteringIteration(filterExpr, res, queryInfo.getStrategy());
+					}
+					if (!res.hasNext()) {
+						Iterations.closeCloseable(res);
+						conn.close();
+						resultHolder.set(new EmptyIteration<>());
+						return;
+					}
+				} else if (bindings.size() > 0) {
+					res = new InsertBindingsIteration(res, bindings);
 				}
-			} else if (bindings.size() > 0) {
-				res = new InsertBindingsIteration(res, bindings);
-			}
 
-			resultHolder.set(new ConsumingIteration(res, federationContext.getConfig().getConsumingIterationMax()));
+				resultHolder.set(new ConsumingIteration(res, federationContext.getConfig().getConsumingIterationMax()));
+			} catch (Throwable t) {
+				res.close();
+				throw t;
+			}
```

Comment:
This is an unrelated issue, but it was causing a test to fail intermittently and I thought I might fix it while I was at it.

The `if (!res.hasNext()) {` line can throw an exception if the iteration was interrupted. To make the code simple and to handle any other cases that might be problematic I've gone ahead and wrapped most of the code that uses the `withConnection(...)` method with try/catch.